### PR TITLE
Update examples, use "nochip" by default, add integration tests, improve test.py, and some bug fixes

### LINKED ISSUE #1114: 4 channel test- Training dataset has fewer elements than batch size
Hi!
I am trying to train on 4 channels (R,G,B, elevation). I am using the master branch in a Docker image with local data.

After many tries I get the same error when the run reach the train command :   'Training dataset has fewer elements than batch size.'
I tried to set batch size to 1 and increase number of epochs, I also tried to both train and validate on image 2 instead of image 3. 
But I get the same error every time. 

Can´t figure if it´s something in my code or the data I have to change?

Message: 

File "/opt/src/rastervision_pytorch_learner/rastervision/pytorch_learner/learner.py", line 541, in setup_data
    'Training dataset has fewer elements than batch size.')

My data:

[https://drive.google.com/drive/folders/1ed0NpcjWOdkiSEuliszkDmytuLqVrdO5?usp=sharing](url)

Image 1

Image 2

Image 3

```
import os
from os.path import join, basename

from rastervision.core.rv_pipeline import *
from rastervision.core.backend import *
from rastervision.core.data import *
from rastervision.core.analyzer import *
from rastervision.pytorch_backend import *
from rastervision.pytorch_learner import *
from rastervision.pytorch_backend.examples.utils import (get_scene_info,
                                                         save_image_crop)
from rastervision.pytorch_backend.examples.semantic_segmentation.utils import (
    example_multiband_transform, example_rgb_transform, `imagenet_stats,
    Unnormalize)

def get_config(runner,
               multiband: bool = True,
               external_model: bool = False,
               augment: bool = False,
               nochip: bool = False,
               test: bool = False):
    root_uri = '/opt/data/output/'
    train_image_uris = ['/opt/data/data_input/images/1.tif','/opt/data/data_input/images/2.tif']
    train_label_uris = ['/opt/data/data_input/labels/1.geojson','/opt/data/data_input/labels/2.geojson']
    train_scene_ids = ['1','2']
    train_scene_list = list(zip(train_scene_ids, train_image_uris, train_label_uris))

    val_image_uri = '/opt/data/data_input/images/3.tif'
    val_label_uri = '/opt/data/data_input/labels/3.geojson'
    val_scene_id = '3'
  

    train_scenes_input = []

    if multiband:
        # use all 4 channels
        channel_order = [0, 1, 2, 3]
        channel_display_groups = {'RGB': (0, 1, 2), 'elev': (3, )}
        aug_transform = example_multiband_transform
    else:
        # use elev, red, & green channels only
        channel_order = [3, 0, 1]
        channel_display_groups = None
        aug_transform = example_rgb_transform

    if augment:
        mu, std = imagenet_stats['mean'], imagenet_stats['std']
        mu, std = mu[channel_order], std[channel_order]

        base_transform = A.Normalize(mean=mu.tolist(), std=std.tolist())
        plot_transform = Unnormalize(mean=mu, std=std)

        aug_transform = A.to_dict(aug_transform)
        base_transform = A.to_dict(base_transform)
        plot_transform = A.to_dict(plot_transform)
    else:
        aug_transform = None
        base_transform = None
        plot_transform = None

    chip_sz = 300
    img_sz = chip_sz
    if nochip:
        chip_options = SemanticSegmentationChipOptions()
    else:
        chip_options = SemanticSegmentationChipOptions(
            window_method=SemanticSegmentationWindowMethod.sliding,
            stride=chip_sz)

    class_config = ClassConfig(
    names=['building', 'background'], colors=['red', 'black'])

    def make_scene(scene_id, image_uri, label_uri):
     
        raster_source = RasterioSourceConfig(
            uris=[image_uri],
            channel_order=channel_order,
            transformers=[StatsTransformerConfig()])
        vector_source = GeoJSONVectorSourceConfig(
            uri=label_uri, default_class_id=0, ignore_crs_field=True)
        label_source = SemanticSegmentationLabelSourceConfig(
            raster_source=RasterizedSourceConfig(
                vector_source=vector_source,
                rasterizer_config=RasterizerConfig(background_class_id=1)))
        return SceneConfig(
            id=scene_id,
            raster_source=raster_source,
            label_source=label_source)


    for scene in train_scene_list:
        train_scenes_input.append(make_scene(*scene))
        
    dataset = DatasetConfig(
    class_config=class_config,
    train_scenes=
        train_scenes_input
    ,
    validation_scenes=[
        make_scene(val_scene_id, val_image_uri, val_label_uri)
    ])
    
    

    # Use the PyTorch backend for the SemanticSegmentation pipeline.
    chip_sz = 300
    backend = PyTorchSemanticSegmentationConfig(
        data=SemanticSegmentationImageDataConfig(),
        model=SemanticSegmentationModelConfig(backbone=Backbone.resnet50),
        solver=SolverConfig(lr=1e-4, num_epochs=10, batch_sz=1, one_cycle=True))
    chip_options = SemanticSegmentationChipOptions(
        window_method=SemanticSegmentationWindowMethod.random_sample,
        chips_per_scene=10)

    return SemanticSegmentationConfig(
        root_uri=root_uri,
        dataset=dataset,
        backend=backend,
        train_chip_sz=chip_sz,
        predict_chip_sz=chip_sz)`
```

### LINKED ISSUE #1117: Learner._get_image_datasets() assumes presence of both 'train' and 'valid' directories in chip zip files
Thanks for catching this @jamesmcclain!

### PR DESCRIPTION
**Pull request by AdeelH**
## Overview

This PR updates all examples in preparation for the 0.13 release. It also includes multiple fixes.

Most of the 'additions' are in `test.py` and its corresponding readme.

### Major changes
- Auto skip `chip` stage when using a `GeoDataConfig`
- Add integration tests for testing `GeoDataConfig`
- Update examples
- Refactor/update `test.py`
  - Add new `compare_runs` command for comparing evals of two runs of an example. See example outputs below.
  - Add new `upload` command for uploading results to model zoo.
  - Add a readme that describes usage

### Minor changes
- Auto skip `analyze` stage when no analyzers are present
- Do not halve `img_sz` in test_mode
  - not really useful
  - causes unnecessary errors
- Add `data/` to .gitignore
  - this is the dir mounted to `/opt/data`

### Fixes
- Propagate `img_channels` to `DataConfig` (#1114)
- Fix bug in `Learner.plot_batch()` (#1114)
- Resize SS prediction output to match input shape
  - to undo any resizing due to img_sz
- Add a small sanity check in `SceneConfig.__repr_args__()`
- Add `rastervision_core` as a dependency for `rastervision_pytorch_learner`
- Do not assume both train and valid dirs are present in each chip zip file (#1117)

## Eval comparisons to 0.12
Diff format is  `new - old`.

### Spacenet Rio
image

### ISPRS Potsdam
image

### Spacenet Vegas - Buildings
image

### Spacenet Vegas - Roads
image

### COWC Potsdam
image

### Xview
image

### Checklist

- [x] Added `needs-backport` label if PR is bug fix that applies to previous minor release
- [x] Ran scripts/format_code and committed any changes
- [x] Documentation updated if needed
- [x] PR has a name that won't get you publicly shamed for vagueness

Fixes #1114 
Fixes #1117 
Makes progress towards #1085

### CODE REVIEW

File: rastervision_pytorch_backend/rastervision/pytorch_backend/examples/tiny_spacenet.py [hunk 1]
```diff
@@ -1,29 +1,28 @@
...
+
     channel_order = [0, 1, 2]
     class_config = ClassConfig(
         names=['building', 'background'], colors=['red', 'black'])
 
-    def make_scene(scene_id, image_uri, label_uri):
+    def make_scene(scene_id: str, image_uri: str,
+                   label_uri: str) -> SceneConfig:
         """
         - StatsTransformer is used to convert uint16 values to uint8.
```
Comment (lewfish) [inline rastervision_pytorch_backend/rastervision/pytorch_backend/examples/tiny_spacenet.py:27]: Should remove this comment about `StatsTransformer`

File: rastervision_pytorch_backend/rastervision/pytorch_backend/examples/tiny_spacenet.py [hunk 2]
```diff
@@ -61,16 +58,19 @@ def make_scene(scene_id, image_uri, label_uri):
     # Use the PyTorch backend for the SemanticSegmentation pipeline.
     chip_sz = 300
     backend = PyTorchSemanticSegmentationConfig(
+        data=SemanticSegmentationGeoDataConfig(
+            scene_dataset=scene_dataset,
+            window_opts=GeoDataWindowConfig(
+                method=GeoDataWindowMethod.random,
+                size=chip_sz,
+                size_lims=(chip_sz, chip_sz + 1),
```
Comment (lewfish) [inline rastervision_pytorch_backend/rastervision/pytorch_backend/examples/tiny_spacenet.py:66]: What is `size_lims` and why is it `chip_sz + 1`?

File: integration_tests/object_detection/config.py [hunk 3]
```diff
@@ -35,7 +37,35 @@ def make_scene(scene_id, img_path, label_path):
...
+                   get_path('scene/labels2.json'))
+    ]
+    scene_dataset = DatasetConfig(
+        class_config=class_config,
+        train_scenes=scenes,
+        validation_scenes=scenes)
+
+    chip_options = ObjectDetectionChipOptions(neg_ratio=1.0, ioa_thresh=1.0)
+
+    if not nochip:
```
Comment (lewfish) [inline integration_tests/object_detection/config.py:54]: You don't have to change this, but double negatives in if statements are a pet peeve.

**Review by lewfish — APPROVED** [review]
Great job! I like the improved example test script and integration tests. The commit structure made it easy to review. I approved it, but there are some small things which you might want to update.

File: rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py [hunk 4]
```diff
@@ -184,6 +187,16 @@ def plot_batch(self,
             constrained_layout=True,
             figsize=(3 * ncols, 3 * nrows))
 
+        # plt.subplots only returns a single AxesSubplot object
+        # if nrows * ncols = 1
+        if not isinstance(axes, np.ndarray):
```
[the remark this answers is not in the document]
Comment (AdeelH) [reply rastervision_pytorch_learner/rastervision/pytorch_learner/semantic_segmentation_learner.py:208]: Ah, that's perfect

File: rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/isprs_potsdam.py [hunk 5]
```diff
@@ -35,9 +35,9 @@ def get_config(runner,
                processed_uri: str,
                root_uri: str,
                multiband: bool = False,
-               external_model: bool = False,
+               external_model: bool = True,
```
[the remark this answers is not in the document]
Comment (AdeelH) [reply rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/isprs_potsdam.py:38]: Oops, didn't mean to

File: rastervision_pytorch_backend/rastervision/pytorch_backend/examples/tiny_spacenet.py (hunk 2 above)
Comment (AdeelH) [reply rastervision_pytorch_backend/rastervision/pytorch_backend/examples/tiny_spacenet.py:66]: Fixed

File: integration_tests/object_detection/config.py (hunk 3 above)
Comment (AdeelH) [reply integration_tests/object_detection/config.py:54]: That's fair

**Review by lewfish — APPROVED** [review]
